### PR TITLE
Vqvae refactor

### PR DESCRIPTION
**Pull request by timonpalm**
Also das ist die beste Variante, die ich hinbekommen habe. Bin eig immer noch nicht ganz mit den Ergebnissen zufrieden.

Ergebnisse für einen 15 Epoch run auf 10.000 imagenet Bildern könnt ihr auf wandb sehen.

Key points:
- es werden jetzt auch wirklich die latents weitergegeben (deswegen natürlich auch schlechtere Ergebnisse)
- ich habe auch ein histogramm für den code use hinzugefügt. Leider gefällt mir das nicht so wie das bei wandb aussieht und hatte da keine lust mehr Zeit reinzustecken. Kann aber auch mit plt geplottet werden

Dinge, die ich noch unklar finde...:
- die testloss ist immer niedriger, als train loss, obwohl visuell schlechter. Deudet eig daraufhin, dass iwas falsch ist :)
- bei meinen tests habe ich gemerkt, dass das initalisieren der codes iwie einen sehr großen Einfluss hat. Gut hat es funktioniert sie zwischen -1 und 1 oder -1/5 und 1/5 zu initalisieren. Ich habe es auch mit 1/K, wie in der einen Implementierung probiert. Das war ganz schlecht. Alle points sind auf einem Punkt und die Idee, dass die latents die codes dann "rausziehen" hat nicht so ganz funktioniert. Bzw. generell bewegen die sich ziemlich wenig. Dachte vllt muss man einfach länger trainieren. Generell sind ja Schwankungen und eine leicht schlechtere Konvergierung glaube ich zu erwarten, da die latents nachm step auf einen anderen code gemapped werden.
- ich habe das standartisieren der Bilder mal weggelassen und im Netz noch sigmoid hinzugefügt. Hab das Gefühl, dass es besser funktioniert hat. Von der Intuition her, haben immer nur relus benutzt und dann den latents auch negative werte zu erlauben, kann glaube schwierig sein dann. Iwie war die auch viel besser. Vllt wenn man alle Relus auf Sigmoids macht dann gehts auch mit standartisieren? hahaha
- Ist häufig auch nicht ganz so gut convergiert. Habe mal nen lr scheduler probiert. Das hat geholfen.

Iwie glaube ich da ist noch was falsch. An so Parameter und Scheduler etc. rumspielen geht iwie immer aber eig sollte mit dem setup, wie es im Paper steht ja wenigsten sinnvolle ergebnisse kommen. 

Wäre natürlich super, wenn ihr da was findet, aber ansonsten sind die ergebnisse auch nicht komplett schrott. Kann man schon mit arbeiten

Man kann die model params auch noch nicht mit dem config anpassen. Sorry dafür aber hab einfach unglaublich viel Zeit in Pixelcnn und das Vqvae gesteckt. Das sind aber die nächsten Steps

Sorry für den Messi text

### CODE REVIEW

File: src/pml_vqvae/stats_keeper.py [hunk 1]
```diff
@@ -74,9 +76,9 @@ def add_batch_stats(self, stats: dict, batch_size: int, train: bool = True):
         if train:
             self.example_cnt += batch_size
 
-        return f"[{'train' if train else 'test'}] " + " | ".join(
-            [f"{k}: {v:.2f}" for k, v in stats.items()]
-        )
+        loss = stats.get("Loss", 0)
+
+        return f"Loss: {loss:.4f}"
```
Comment (rosinusserrano) [inline src/pml_vqvae/stats_keeper.py:82]: warum nicht mehr alle stats printen?

File: src/pml_vqvae/vqvae/vqvae.py [hunk 2]
```diff
@@ -186,26 +187,29 @@ def backward(self, loss: torch.Tensor):
...
+            plt.legend()
 
         # set window
-        plt.xlim(-2, 2)
-        plt.ylim(-2, 2)
+        plt.xlim(-4, 4)
+        plt.ylim(-4, 4)
 
         plt.title("Codes")
         plt.xlabel("x1")
```
Comment (rosinusserrano) [inline src/pml_vqvae/vqvae/vqvae.py:213]: kann man dafür nicht die die visualize output funktion benutzen?

File: src/pml_vqvae/vqvae/vqvae.py [hunk 3]
```diff
@@ -111,7 +109,7 @@ def __init__(
                 stride=2,
                 padding=1,
             ),
-            # torch.nn.Tanh()
+            torch.nn.Sigmoid(),
```
Comment (rosinusserrano) [inline src/pml_vqvae/vqvae/vqvae.py]: Hat das den unterschied gemacht oder kann man das zurückändern?

File: src/pml_vqvae/vqvae/vqvae.py [hunk 4]
```diff
@@ -222,5 +226,24 @@ def visualize_output(batch, output, target, prefix: str = "", base_dir: str = ".
...
+
+        if self.prior is None:
+            raise ValueError("No prior set. Use set_prior() to set a prior model")
+
+        latents = self.prior.sample(class_idx)
+
+        output = self.decoder_stack(latents)
+
+        return output
+
```
Comment (rosinusserrano) [inline src/pml_vqvae/vqvae/vqvae.py:245]: ich würd das nicht in das model integrieren, benutzen ja das gleiche VQVAE für alle klassen. Glaube ne funktion, oder ne klasse, die nen pixelcnn und n vqvae als input nimmt, und dann daraus bilder samplet sollte genügen. Wir benutzen die funktion ja nicht n der training pipeline

File: src/pml_vqvae/stats_keeper.py (hunk 1 above)
Comment (timonpalm) [reply src/pml_vqvae/stats_keeper.py:82]: Hab da noch die Codebase coverage mit reingemacht. Die konnte man nicht wirklich printen

File: src/pml_vqvae/vqvae/vqvae.py (hunk 2 above)
Comment (timonpalm) [reply src/pml_vqvae/vqvae/vqvae.py:213]: Ist ja basically das. visualise output wird auch nirgends benutzt. Die Funktion geht nur, wenn die latent 2 dimensional sind

File: src/pml_vqvae/vqvae/vqvae.py (hunk 4 above)
Comment (timonpalm) [reply src/pml_vqvae/vqvae/vqvae.py:245]: Geschmacks Sache. Beides möglich
